src/core/processing/models/qgsprocessingmodelalgorithm.cpp

```diff
@@ -580,6 +580,11 @@ QMap<QString, QgsProcessingModelAlgorithm::VariableDefinition> QgsProcessingMode
       QgsProcessingFeatureSourceDefinition fromVar = qvariant_cast<QgsProcessingFeatureSourceDefinition>( value );
       value = fromVar.source;
     }
+    else if ( value.canConvert<QgsProcessingOutputLayerDefinition>() )
+    {
+      QgsProcessingOutputLayerDefinition fromVar = qvariant_cast<QgsProcessingOutputLayerDefinition>( value );
+      value = fromVar.sink;
```

Comment:
This does not consider that list of Outvariable (e.g. i raster calculator) are filled with a list of:
"Output '%1' from algorithm '%2'"  with wrong name (e.g. Output in %1) as in line:
https://github.com/nyalldawson/QGIS/blob/8c81c513dbbbe6511ccd8036719e7c56c0936c4a/src/core/processing/models/qgsprocessingmodelalgorithm.cpp#L516
and
https://github.com/nyalldawson/QGIS/blob/8c81c513dbbbe6511ccd8036719e7c56c0936c4a/src/core/processing/models/qgsprocessingmodelalgorithm.cpp#L470

Reply:
Sorry, I don't follow?

Reply:
try, create a model with two raster cascaded as in attached model  (need renamed to .model3).
[rastercalculator_cascaded.txt](https://github.com/qgis/QGIS/files/2179501/rastercalculator_cascaded.txt)

you can find that list of possible input variables of a calculator that comes from a previous algorithm has always the name "Output 'Output' from algorithm '<thealgid>' " instead of:
"Output '<sink name>' from algorithm '<thealgid>' " if a sink has been specified.

Reply:
Looks ok here:

![image](https://user-images.githubusercontent.com/1829991/42499196-83317032-8471-11e8-8e9d-4aebc262afd6.png)

Reply:
shouldn't it be RESULT1 (the sink destinationName) instead of "Output"?
e.g.
"RESULT1 '' from algorithm ''calc1"
instead of
"Output '' from algorithm ''calc1"

Reply:
It's using the name of the corresponding output from the algorithm. E.g. expand out the + under the algorithm in the modeler and you'll see the available outputs and their names. But most algorithms use the convention of "output" as the output name for generated layers.

Reply:
ok, I'm trying to find a way to add in the expressionContext a reliable mapping of input vars with that added in the rasterCalculator expression. btw this is out of the scope of this PR. forgot my latest comments.

Reply:
@luipir I'll weigh in on that when I'm back fully online at the end of the week